basil/HL/tektronix_oscilloscope.py

```diff
@@ -0,0 +1,197 @@
+#
+
+from basil.HL.scpi import scpi
+from collections import namedtuple
+from enum import Enum
+
+Identity = namedtuple("Identity", "company, model, serial, config")
+XScale = namedtuple("XScale", "slope, offset, unit")
+YScale = namedtuple("YScale", "top, bottom")
+FeatureTable = namedtuple("FeatureTable", "name, entries")
+Waveform = namedtuple("Waveform", "data, x_scale, y_scale")
+
+
+class WaveType(Enum):
+    ANALOG = 1
+    DIGITAL = 2
+    MATH = 3
+
+
+JobParameters = namedtuple(
+    "JobParameters",
+    ["wave_type", "channel_name", "encoding", "bit_nr", "data_type", "record_length"],
+)
+
+
+class WaveformCollection:
+    def __init__(self):
+        self.idn = None
+        self.data = {}
+
+    @property
+    def sources(self):
+        return list(self.data.keys())
+
+    def __getitem__(self, item):
+        return self.data[item]
+
+    def __len__(self):
+        return len(self.data.keys())
+
+
+class TektronixOscilloscope(scpi):
+
+    def _make_jobs(self, channel=1):
+        """Scan the instrument for available sources of data construct a list of jobs"""
+
+        encoding_table = {
+            WaveType.MATH: ("FPBinary", 16, "f"),
+            WaveType.DIGITAL: ("RIBinary", 16, "h"),
+            WaveType.ANALOG: ("RIBinary", 16, "h"),
+        }
+        channel_table = {
+            WaveType.MATH: lambda x: x,
+            WaveType.DIGITAL: lambda x: "_".join([x.split("_")[0], "DALL"]),
+            WaveType.ANALOG: lambda x: x,
+        }
+
+        results = {}
+        sources = ['CH' + str(channel)]
+        for source in sources:
+            if source.split("_")[0] not in results:
+                # Determine the type of waveform and set key interface parameters
+                wave_type = self._classify_waveform(source)
+                channel = channel_table[wave_type](source)
+                encoding, bit_nr, datatype = encoding_table[wave_type]
+
+                # Set the start and stop point of the record
+                rec_len = int(self._intf.query("horizontal:recordlength?").strip())
+
+                # Keep track of each super channel and math source that has been handled
+                results[source.split("_")[0]] = JobParameters(
+                    wave_type, channel, encoding, bit_nr, datatype, rec_len
+                )
+        return results
+
+    def _classify_waveform(self, source):
+        if source[0:4] == "MATH":
+            wave_type = WaveType.MATH
+        elif "_" in source:
+            wave_type = WaveType.DIGITAL
+        else:
+            wave_type = WaveType.ANALOG
+        return wave_type
+
+    def _get_xscale(self):
+        """Get the horizontal scale of the instrument"""
+
+        # This function will generate a VISA timeout if there is no waveform data
+        # available and the channel is enabled.
+        result = None
+        try:
+            xincr = self._intf.query("WFMOutpre:XINCR?").strip()
+        except Exception:
```

Comment:
No bare `except` statement. Please catch the timeout generated by pyvisa, otherwise hard to debug if things are not working